gdpr/ipcypher.py

```diff
@@ -0,0 +1,178 @@
+"""
+Implementation of ipcipher.
+
+https://github.com/PowerDNS/ipcipher
+
+IPv4 took from https://github.com/veorq/ipcrypt/blob/939549e3f542c7ae9eb1aec96164f19b5f9fc46c/ipcrypt.py under CC0
+"""
+
+# type: ignore
+# flake8: noqa
+from hashlib import pbkdf2_hmac
+from ipaddress import IPv4Address, IPv6Address, ip_address
+from typing import Union
+
+from pyaes import AESModeOfOperationECB
+
+
+__all__ = ['derive_key', 'encrypt_ipv4', 'decrypt_ipv4', 'encrypt_ipv6', 'decrypt_ipv6', 'encrypt_ip', 'decrypt_ip']
+
+IPv4Type = Union[str, IPv4Address]
+IPv6Type = Union[str, IPv6Address]
+IPType = Union[IPv4Type, IPv6Type]
+
+
+def derive_key(key: str):
+    """
+    PBKDF2(SHA1, Password, "ipcipheripcipher", 50000, 16)
+    """
+
+    return pbkdf2_hmac('sha1', bytes(key, encoding="utf-8"), bytes('ipcipheripcipher', encoding='utf-8'), 50000, 16)
+
+
+def rotl(b, r):
+    return ((b << r) & 0xff) | (b >> (8 - r))
+
+
+def permute_fwd(state):
```

Comment:
také prosím zkus popsat proč to permutuješ

Reply:
nebo zda jsi to z něčeho nevzal. Je to tak tricky, že si nejsem jist zda jsi to vymýšlel. Převzaté věci jsou dobré jen je dobré vědět odkud

Reply:
ano myslel. Možná jsem to přehlédl, možná bylo doplněno. Nevím